ocaml-lsp-server/src/compl.ml

```diff
@@ -79,42 +71,42 @@ let prefix_of_position ~short_path source position =
         | ':'
         | '~'
         | '#' ->
-          find (ch :: prefix) (i - 1)
-        | '.' ->
-          if short_path then
-            make_string prefix
-          else
-            find (ch :: prefix) (i - 1)
-        | _ -> make_string prefix
+          false
+        | '.' -> short_path
+        | _ -> true)
     in
-
-    let (`Offset index) = Msource.get_offset source position in
-    find [] (index - 1)
+    let pos =
+      match pos with
+      | None -> 0
+      | Some pos -> pos + 1
+    in
+    let len = from - pos + 1 in
+    String.sub text ~pos ~len
 
 let suffix_of_position source position =
   match Msource.text source with
   | "" -> ""
   | text ->
     let len = String.length text in
-
-    let rec find suffix i =
-      if i >= len then
-        suffix
-      else
-        let ch = text.[i] in
-        (* The characters for an infix function are missing *)
-        match ch with
-        | 'a' .. 'z'
-        | 'A' .. 'Z'
-        | '0' .. '9'
-        | '\''
-        | '_' ->
-          find (ch :: suffix) (i + 1)
-        | _ -> suffix
+    let from =
+      let (`Offset index) = Msource.get_offset source position in
+      min index (len - 1)
     in
-
-    let (`Offset index) = Msource.get_offset source position in
-    make_string (List.rev @@ find [] index)
+    let len =
+      let until =
+        String.findi ~from text ~f:(function
+          | 'a' .. 'z'
+          | 'A' .. 'Z'
+          | '0' .. '9'
+          | '\''
+          | '_' ->
+            false
+          | _ -> true)
```

Comment:
I think it would be very helpful in reading the code if we moved predicates out in a separate function something like `is_valid_ident` (in our case it's more like `not_is_valid_ident`) here and in `prefix_of_position`.

Reply:
That's true. I did that.